# Segmenting and Clustering Neighbourhoods in Toronto
### Parts 1 to 3

The goal of this work is performing an analysis and clustering of the different neighborhoods in the city Toronto according to a certain criteria.

* In [Part 1](#Part-1) the Postal Code data of Canada is obtained from wikipedia.
* In [Part 2](#Part-2) the data from postal codes and its geographical coordinates are merged with the neighborhood data.
* In [Part 3](#Part-3) the merged data is used along with the Foursquare API data in order to set up the criteria for segmenting and clustering the neighborhoods. 

Please note the maps can't be shown on Github, the URL to this notebook is on the README!

In [1]:
## loading libraries 
import requests
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
!pip install folium
import folium # map rendering library
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans # import k-means from clustering stage

<a id='Part_1'></a>

# Part 1

Scraping data from Wikipedia (https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) using pandas to a dataframe. 
* Cells with Borough equals to **"Not assigned"** are ignored.
* Neighborhoods within the same borough (and Postal Code) are grouped and separated by commas

In [2]:
postCodeList = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header=0)[0]
df = postCodeList[~postCodeList["Borough"].str.contains("Not assigned")]
df = df.replace('/', ',', regex=True)

df_canada = df.sort_values(by=["Postal Code"])
df_canada.set_index("Postal Code")
df_canada.head()

Postal Code      Borough                            Neighborhood
9          M1B  Scarborough                          Malvern, Rouge
18         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
27         M1E  Scarborough       Guildwood, Morningside, West Hill
36         M1G  Scarborough                                  Woburn
45         M1H  Scarborough                               Cedarbrae

In [3]:
print(df_canada.shape)

(103, 3)


The dataframe has 103 different rows.

<a id='Part_2'></a>
# Part 2

Obtaining the .csv file from http://cocl.us/Geospatial_data:
* It generates a dataframe with Postal Code, Latitude and Longitude
* Merges with the previous dataframe to gather all the info in a single dataframe.

In [4]:
dflatlon= pd.read_csv("http://cocl.us/Geospatial_data")
dflatlon.set_index("Postal Code")
dflatlon.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Merging the tables according to Postal Code

In [5]:
df_canada_full = pd.merge(df_canada,dflatlon, on="Postal Code")
df_canada_full.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

<a id='Part_3'></a>
# Part 3

Now the analysis and clustering of the previously obtained data using Foursquare API data

#### Functions and API Keys

In [6]:
## API Key hidden due to security reasons


In [7]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Separating the data to use only the Boroughs whose names contain "Toronto"

In [8]:
toronto_data = df_canada_full[df_canada_full['Borough'].str.contains("Toronto")]
toronto_data.head()

Postal Code          Borough                    Neighborhood   Latitude  \
37         M4E     East Toronto                     The Beaches  43.676357   
41         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43         M4M     East Toronto                 Studio District  43.659526   
44         M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

#### Using geocoders library to obtain Toronto geographical coordinates

In [9]:
address = 'Toronto, CA'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


#### Building a map with the neighborhoods (grouped by Post Code) in Toronto area

In [10]:
# create map using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Obtaining all the narby venues in the neighborhoods  using the Foursquare API

In [11]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],latitudes=toronto_data['Latitude'],longitudes=toronto_data['Longitude'])
toronto_venues.head()

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

### Counting of the data according to neighborhood groups 

In [12]:
toronto_venues.groupby("Neighborhood").count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     18   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     64   
Christie                                                               17   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          22   
India Bazaar, The Beaches West                                         23   
Kensington Market, Chinatown, Grange Park                              57   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             3   
North Toronto West, Lawrence Park                                      20   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                               92   
Rosedale                                                                4   
Roselawn                                                                4   
Runnymede, Swansea                                                     41   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            47   
Stn A PO Boxes                                                         92   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      18   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      64   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                    

In [13]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))


There are 237 unique categories.


In [14]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'] )
toronto_onehot.shape

(1616, 237)

In [15]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

toronto_grouped.shape

(39, 237)

#### Obtaining the top 6 venues for each neighborhood group 

In [16]:
num_top_venues = 6

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04
5            Beer Bar  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3   Grocery Store  0.05
4    Intersection  0.05
5   Burrito Place  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3         Pizza Place  0.06
4          Comic Shop  0.06
5    Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3             Plane  0.06
4           Airport  0.

               venue  freq
0     Sandwich Place  0.13
1               Café  0.13
2        Coffee Shop  0.09
3      Grocery Store  0.04
4  Indian Restaurant  0.04
5     Cosmetics Shop  0.04


----The Beaches----
                             venue  freq
0                            Trail  0.25
1                Health Food Store  0.25
2                              Pub  0.25
3                Afghan Restaurant  0.00
4              Monument / Landmark  0.00
5  Molecular Gastronomy Restaurant  0.00


----The Danforth West, Riverdale----
                    venue  freq
0        Greek Restaurant  0.21
1             Coffee Shop  0.07
2      Italian Restaurant  0.07
3          Ice Cream Shop  0.05
4              Restaurant  0.05
5  Furniture / Home Store  0.05


----Toronto Dominion Centre, Design Exchange----
                 venue  freq
0          Coffee Shop  0.10
1                 Café  0.06
2                Hotel  0.06
3   Seafood Restaurant  0.03
4  American Restaurant  0.03
5          Sal

In [17]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Sorting the data and resume

In [18]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
toronto_venues_sorted = pd.DataFrame(columns=columns)
toronto_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

toronto_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                  Café              Beer Bar   
1        Breakfast Spot           Coffee Shop         Burrito Place   
2         Auto Workshop            Skate Park               Brewery   
3       Airport Service      Airport Terminal      Sculpture Garden   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant           Cheese Shop   
1    Italian Restaurant            Restaurant          Intersection   
2         Burrito Place                   Spa        Farmers Market   
3       Harbor / Marina   Rental Car Location                 Plane   
4   Japanese Restaurant        Ice Cream Shop           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant                   Pub         Farmers Market  
1               Stadium                   Bar                 Bakery  
2  Fast Food Restaurant            Restaurant       Recording Studio  
3           Coffee Shop         Boat or Ferry                    Bar  
4       Bubble Tea Shop       Thai Restaurant           Burger Joint

#### Classificating the neighborhoods into 5 clusters 

In [19]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [20]:
# add clustering labels
toronto_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code          Borough                    Neighborhood   Latitude  \
37         M4E     East Toronto                     The Beaches  43.676357   
41         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43         M4M     East Toronto                 Studio District  43.659526   
44         M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               0                 Trail                   Pub   
41 -79.352188               1      Greek Restaurant           Coffee Shop   
42 -79.315572               1                  Park        Sandwich Place   
43 -79.340923               1                  Café           Coffee Shop   
44 -79.388790               2                  Park              Bus Line   

   3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
37     Health Food Store               Dog Run        Department Store   
41    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
42  Fast Food Restaurant     Food & Drink Shop            Liquor Store   
43             Gastropub                Bakery                 Brewery   
44           Swim School           Event Space       Electronics Store   

          6th Most Common Venue 7th Most Common Venue  \
37                 Dessert Shop                 Diner   
41                   Restaurant       Bubble Tea Shop   
42                Burrito Place    Italian Restaurant   
43          American Restaurant           Yoga Studio   
44  Eastern European Restaurant   Dumpling Restaurant   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
37           Discount Store   Distribution Center            Yoga Studio  
41                   Bakery                   Pub            Pizza Place  
42               Restaurant          Intersection         Ice Cream Shop  
43  Comfort Food Restaurant    Seafood Restaurant         Sandwich Place  
44               Donut Shop      Doner Restaurant                Dog Run

#### Clustering result into map

In [21]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining Clusters


In [22]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37  East Toronto               0                 Trail                   Pub   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37     Health Food Store               Dog Run      Department Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37          Dessert Shop                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
37   Distribution Center            Yoga Studio

In [23]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
41      East Toronto               1      Greek Restaurant   
42      East Toronto               1                  Park   
43      East Toronto               1                  Café   
45   Central Toronto               1                  Park   
46   Central Toronto               1           Coffee Shop   
47   Central Toronto               1          Dessert Shop   
49   Central Toronto               1                   Pub   
51  Downtown Toronto               1           Coffee Shop   
52  Downtown Toronto               1           Coffee Shop   
53  Downtown Toronto               1           Coffee Shop   
54  Downtown Toronto               1        Clothing Store   
55  Downtown Toronto               1                  Café   
56  Downtown Toronto               1           Coffee Shop   
57  Downtown Toronto               1           Coffee Shop   
58  Downtown Toronto               1           Coffee Shop   
59  Downtown Toronto               1           Coffee Shop   
60  Downtown Toronto               1           Coffee Shop   
61  Downtown Toronto               1           Coffee Shop   
65   Central Toronto               1        Sandwich Place   
66  Downtown Toronto               1                  Café   
67  Downtown Toronto               1                  Café   
68  Downtown Toronto               1        Airport Lounge   
69  Downtown Toronto               1           Coffee Shop   
70  Downtown Toronto               1           Coffee Shop   
75  Downtown Toronto               1         Grocery Store   
76      West Toronto               1                Bakery   
77      West Toronto               1                   Bar   
78      West Toronto               1                  Café   
82      West Toronto               1       Thai Restaurant   
83      West Toronto               1             Gift Shop   
84      West Toronto               1           Pizza Place   
85  Downtown Toronto               1           Coffee Shop   
87      East Toronto               1    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue        4th Most Common Venue  \
41           Coffee Shop     Italian Restaurant               Ice Cream Shop   
42        Sandwich Place   Fast Food Restaurant            Food & Drink Shop   
43           Coffee Shop              Gastropub                       Bakery   
45        Sandwich Place       Department Store                        Hotel   
46        Clothing Store            Yoga Studio                         Café   
47        Sandwich Place                    Gym             Sushi Restaurant   
49           Coffee Shop            Supermarket        Vietnamese Restaurant   
51           Pizza Place                    Pub                       Bakery   
52   Japanese Restaurant       Sushi Restaurant                   Restaurant   
53                  Park                 Bakery                          Pub   
54           Coffee Shop             Restaurant                         Café   
55           Coffee Shop           Cocktail Bar                    Gastropub   
56          Cocktail Bar                   Café                     Beer Bar   
57    Italian Restaurant         Sandwich Place                         Café   
58                  Café             Restaurant                Deli / Bodega   
59              Aquarium                  Hotel                         Café   
60                 Hotel                   Café           Seafood Restaurant   
61                  Café             Restaurant                        Hotel   
65                  Café            Coffee Shop               Cosmetics Shop   
66            Restaurant                    Bar           Italian Restaurant   
67                Bakery  Vietnamese Restaurant           Mexican Restaurant   
68       Airport Service       Airport Terminal             Sculpture Garden   
69                  Café           Cocktail Bar           Italia

In [24]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
44   Central Toronto               2                  Park   
50  Downtown Toronto               2                  Park   
64   Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
44              Bus Line           Swim School           Event Space   
50            Playground                 Trail          Cycle Studio   
64         Jewelry Store                 Trail              Bus Line   

          5th Most Common Venue        6th Most Common Venue  \
44            Electronics Store  Eastern European Restaurant   
50  Eastern European Restaurant          Dumpling Restaurant   
64             Sushi Restaurant                  Yoga Studio   

   7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
44   Dumpling Restaurant                   Donut Shop      Doner Restaurant   
50            Donut Shop             Doner Restaurant               Dog Run   
64      Department Store  Eastern European Restaurant   Dumpling Restaurant   

   10th Most Common Venue  
44                Dog Run  
50    Distribution Center  
64             Donut Shop

In [25]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
48  Central Toronto               3            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
48                 Trail          Intersection           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
48         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
48   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [26]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
63  Central Toronto               4                Garden   

   2nd Most Common Venue    3rd Most Common Venue 4th Most Common Venue  \
63           Music Venue  Health & Beauty Service        Ice Cream Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
63           Yoga Studio      Department Store          Dessert Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
63                 Diner        Discount Store                Dog Run